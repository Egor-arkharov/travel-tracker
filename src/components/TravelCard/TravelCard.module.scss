.card {
  height: 100%;
  background: #fff;
  border-radius: 12px;
  overflow: hidden;
  box-shadow: 0 4px 12px rgba(0, 0, 0, 0.08);
  display: flex;
  flex-direction: column;
  transition: transform 0.2s;

  &:hover {
    transform: translateY(-4px);
  }

  container-type: inline-size;
}

.inner {
  height: 100%;
  background: #fff;

  @container (min-width: 500px) {
    display: flex;

    .imageWrapper {
      width: 50%;
    }

    .image {
      border-radius: 12px;
    }
  }
}

.imageWrapper {
  position: relative;
  width: 100%;
  height: 200px;
}

.image {
  object-fit: cover;
  width: 100%;
  height: 100%;
}

.info {
  padding: 16px;
  font-size: 0.95rem;
  line-height: 1.4;
}
