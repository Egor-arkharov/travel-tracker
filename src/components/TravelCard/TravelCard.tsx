// components/TravelCard.tsx
import Image from "next/image";
import { Travel } from "@/types/travel";
import styles from "./TravelCard.module.scss";

const TravelCard = ({ travel }: { travel: Travel }) => {
  return (
    <div className={styles.card}>
      <div className={styles.inner}>
        {travel.imageUrl && (
          <div className={styles.imageWrapper}>
            <Image
              src={travel.imageUrl}
              alt={`${travel.city} photo`}
              width={400}
              height={300}
              className={styles.image}
            />
          </div>
        )}
        <div className={styles.info}>
          <h2>{travel.country}</h2>
          <p>City: {travel.city}</p>
          <p>Date: {travel.startDate} â€“ {travel.endDate}</p>
          <p>Budget: ${travel.budget}</p>
          <p>Rating: {travel.rating}</p>
        </div>
      </div>

    </div>
  );
};

export default TravelCard;
