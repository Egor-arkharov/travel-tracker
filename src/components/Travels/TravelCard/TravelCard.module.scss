// Travels/TravelCard/TravelCard.module.scss
@use "@/styles/global" as *;

.card {
  height: 100%;
  background: #fff;
  overflow: hidden;
  box-shadow: 0 4px 12px rgba(0, 0, 0, 0.08);
  display: flex;
  flex-direction: column;
  cursor: pointer;

  container-type: size;
}


.inner {
  display: flex;
  height: 100%;
  background: white;
  border-radius: 12px;

  .grid & {
    @include media-min(xl) {

      // Высокая карточка
      @container (min-height: 301px) {
        flex-direction: column;

        .imageWrapper {
          width: 100%;
          height: 60%;
        }
      }

      // Широкая карточка
      @container (min-width: 494px) {
        flex-direction: row;

        .imageWrapper {
          width: 65%;
          height: 100%;
        }
      }
    }

    @include media-between(lg, xl) {

      // Обычная карточка
      @container (max-height: 351px) {
        flex-direction: column;

        .imageWrapper {
          width: 100%;
          height: 50%;
        }
      }

      // Высокая карточка
      @container (min-height: 351px) {
        flex-direction: column;

        .imageWrapper {
          width: 100%;
          height: 70%;
        }
      }

      // Широкая карточка
      @container (min-width: 700px) {
        flex-direction: row;

        .imageWrapper {
          width: 65%;
          height: 100%;
        }
      }
    }

    @include media-between(md, lg) {

      // Обычная карточка
      @container (max-height: 301px) {
        flex-direction: column;

        .imageWrapper {
          width: 100%;
          height: 40%;
        }
      }

      // Высокая карточка
      @container (min-height: 301px) {
        flex-direction: column;

        .imageWrapper {
          width: 100%;
          height: 60%;
        }
      }

      // Широкая карточка
      @container (min-width: 570px) {
        flex-direction: row;

        .imageWrapper {
          width: 55%;
          height: 100%;
        }
      }
    }

    @include media-between(sm, md) {

      // Обычная карточка
      @container (max-width: 430px) {
        flex-direction: column;

        .imageWrapper {
          width: 100%;
          height: 50%;
        }
      }

      // Широкая карточка
      @container (min-width: 431px) {
        flex-direction: row;

        .imageWrapper {
          width: 60%;
          height: 100%;
        }
      }
    }

    @include media-between(xs, sm) {

      // Обычная карточка
      @container (max-width: 430px) {
        flex-direction: column;

        .imageWrapper {
          width: 100%;
          height: 40%;
        }
      }

      // Широкая карточка
      @container (min-width: 431px) {
        flex-direction: row;

        .imageWrapper {
          width: 50%;
          height: 100%;
        }
      }
    }
  }

  .list & {
    display: flex;
  }
}

.imageWrapper {
  position: relative;
  overflow: hidden;
  width: 50%;
  flex-shrink: 0;
  border-radius: 12px;

  [data-mode="modal"] & {
    width: auto;
    flex-grow: 1;
    flex-shrink: unset;
  }

  .image {
    display: block;
    width: 100%;
    height: 100%;
    object-fit: cover;

    will-change: transform;
    backface-visibility: hidden;
  }

  @media (width <=500px) {
    order: var(--image-order, 0);
  }
}

.info {
  padding: 16px 12px;

  flex-grow: 1;
  display: flex;
  flex-direction: column;
  gap: 4px;

  color: black;

  h3 {
    margin-bottom: 6px;

    [data-mode="modal"] & {
      font-size: 42px;
    }
  }

  [data-mode="modal"] & {
    gap: 8px;
  }
}

.infoItem {
  display: flex;
  align-items: center;
  gap: 7px;

  [data-mode="modal"] & {
    font-size: 18px;
  }
}

.infoDate {
  white-space: nowrap;

  @media (width <=500px) {
    white-space: normal;
  }
}

.actions {
  display: flex;
  gap: 10px;
  margin-top: 20px;
}

.actions button {
  display: flex;
  align-items: center;
  justify-content: center;

  margin: 0;
  padding: 15px 30px;

  font-family: var(--font-body);
  font-size: 20px;
  font-weight: 700;

  text-decoration: none;

  border: none;
  border-radius: 12px;
  outline: none;

  cursor: pointer;

  &.actionsEdit {
    color: black;
    background-color: var(--blue-soft);

    &:hover {
      background-color: var(--blue-light);
    }

    &:active {
      background-color: var(--blue-dark);
    }
  }

  &.actionsRemove {
    color: black;
    background-color: var(--red-soft);

    &:hover {
      background-color: var(--red-light);
    }

    &:active {
      background-color: var(--red-dark);
    }
  }
}