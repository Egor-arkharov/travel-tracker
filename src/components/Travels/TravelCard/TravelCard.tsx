// Travels/TravelCard/TravelCard.tsx
"use client";

import { Travel } from "@/types/travel";
import styles from "./TravelCard.module.scss";
import TravelImage from "./TravelCardImage";
import TravelInfo from "./TravelCardInfo";
import { motion } from "framer-motion";

interface TravelCardProps {
  travel: Travel;
  view: string;
  onClick: () => void;
  cardLayoutId: string;
  imageLayoutId: string;
  isSelected: boolean;
}

const TravelInfoMotion = motion.create(TravelInfo);

const TravelCard = ({
  travel,
  view,
  onClick,
  cardLayoutId,
  imageLayoutId,
  isSelected,
}: TravelCardProps) => {
  return (
    <motion.article
      layoutId={cardLayoutId}
      className={`${styles.card} ${styles[view]}`}
      onClick={onClick}
      transition={{ type: "spring", stiffness: 400, damping: 35 }}
    >
      <div className={styles.inner}>
        <TravelImage
          travel={travel}
          layoutId={imageLayoutId}
          priority={isSelected}
        />
        <TravelInfoMotion
          travel={travel}
          mode="card"
          initial={false}
          animate={isSelected ? "hidden" : "visible"}
          variants={{
            visible: { opacity: 1, y: 0, transition: { duration: 0.18 } },
            hidden:  { opacity: 0, y: -8, transition: { duration: 0.12 } },
          }}
        />
      </div>
    </motion.article>
  );
};

export default TravelCard;
