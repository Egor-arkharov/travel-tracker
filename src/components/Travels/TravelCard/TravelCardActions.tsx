// TravelCardActions.tsx

import Link from "next/link";
import styles from "./TravelCard.module.scss";
import { Travel } from "@/types/travel";

interface Props {
  travel: Travel;
  onDelete?: () => void;
}

const TravelCardActions = ({ travel, onDelete }: Props) => {
  return (
    <div className={styles.actions}>
      <Link
        href={`/edit/${travel.id}`}
        className="button button--triary"
        aria-disabled={travel.meta.isMock}
      >
        Edit
      </Link>
      <button
        className="button button--danger"
        disabled={travel.meta.isMock}
        onClick={() => {
          if (!travel.meta.isMock && onDelete) onDelete();
        }}
      >
        Remove
      </button>
    </div>
  );
};

export default TravelCardActions;
