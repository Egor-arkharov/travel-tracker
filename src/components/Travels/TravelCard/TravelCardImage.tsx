// Travels/TravelCard/TravelCardImage.tsx
"use client";
import { motion } from "framer-motion";
import Image from "next/image";
import styles from "./TravelCard.module.scss";
import { Travel } from "@/types/travel";

const TravelImage = ({
  travel,
  layoutId,
}: {
  travel: Travel;
  layoutId: string;
}) => {
  const imageSrc = travel.meta.isMock
    ? travel.media.imagePath
    : travel.media.imageUrl;

  if (!imageSrc) return null;

  return (
    <motion.div
      layoutId={layoutId}
      className={styles.imageWrapper}
    >
      <Image
        src={imageSrc}
        alt={travel.location.city}
        fill
        sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
        className={styles.image}
      />
    </motion.div>
  );
};

export default TravelImage;