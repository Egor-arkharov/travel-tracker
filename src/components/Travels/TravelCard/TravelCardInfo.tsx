// Travels/TravelCard/TravelCardInfo.tsx
"use client";

import {
  CountryIcon,
  CalendarIcon,
  BudgetIcon,
  RatingIcon,
  DescriptionIcon,
} from "@/components/icons";

import styles from "./TravelCard.module.scss";
import { Travel } from "@/types/travel";
import { forwardRef } from "react";

interface TravelInfoProps {
  travel: Travel;
  mode?: "card" | "modal";
  onDelete?: () => void;
}

const TravelInfo = forwardRef<HTMLDivElement, TravelInfoProps>(
  ({ travel, mode = "card", onDelete }, ref) => {
    const isModal = mode === "modal";

    return (
      <div className={styles.info} ref={ref}>
        <h3>{travel.location.country}</h3>
        <p className={styles.infoItem}>
          <CountryIcon className={styles.icon} width={14} height={14} />
          <span>{travel.location.city}</span>
        </p>
        <p className={styles.infoItem}>
          <CalendarIcon className={styles.icon} width={14} height={14} />
          <span>
            <span className={styles.infoDate}>{travel.dates.start}</span> â€“{" "}
            <span className={styles.infoDate}>{travel.dates.end}</span>
          </span>
        </p>
        <p className={styles.infoItem}>
          <BudgetIcon className={styles.icon} width={14} height={14} />
          <span>{travel.budget}</span>
        </p>
        <p className={styles.infoItem}>
          <RatingIcon className={styles.icon} width={14} height={14} />
          <span>{travel.rating}</span>
        </p>

        {isModal && (
          <>
            {travel.description && (
              <p className={`${styles.infoDescription} ${styles.infoItem}`}>
                <DescriptionIcon className={styles.icon} width={14} height={14} />
                <span>{travel.description}</span>
              </p>
            )}
            <div className={styles.actions}>
              <button
                className={styles.actionsEdit}
                disabled={travel.meta.isMock}
              >
                Edit
              </button>
              <button
                className={styles.actionsRemove}
                disabled={travel.meta.isMock}
                onClick={() => {
                if (!travel.meta.isMock && onDelete) onDelete();
              }}
              >
                Remove
              </button>
            </div>
          </>
        )}
      </div>
    );
  }
);

TravelInfo.displayName = "TravelInfo";

export default TravelInfo;
