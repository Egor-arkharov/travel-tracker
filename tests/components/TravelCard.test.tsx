import { render } from "@testing-library/react";
import TravelCard from "@/components/Travels/TravelCard/TravelCard";
import type { Travel } from "@/types/travel";

// ðŸ§© Firebase/Auth mocks
jest.mock("@/hooks/useAuth", () => ({
  useAuth: () => ({ isLoggedIn: true, login: jest.fn(), logout: jest.fn() }),
}));
jest.mock("firebase/auth", () => ({}));
jest.mock("@/app/firebase", () => ({}));

// ðŸ§© Mock nested subcomponents
jest.mock("@/components/Travels/TravelCard/TravelCardInfo", () => ({
  __esModule: true,
  default: () => <div data-testid="mock-info" />,
}));
jest.mock("@/components/Travels/TravelCard/TravelCardImage", () => ({
  __esModule: true,
  default: () => <div data-testid="mock-image" />,
}));


const mockTrip: Travel = {
  id: "test-id",
  location: { city: "Athens", country: "Greece" },
  dates: { start: "2024-07-05", end: "2024-07-10" },
  budget: 1200,
  rating: 4,
  media: { imagePath: "", imageUrl: "/images/mock.jpg" },
  meta: { isMock: false },
  description: "A summer trip to Greece",
};

describe("TravelCard", () => {
  test("matches snapshot", () => {
    const { asFragment } = render(
      <TravelCard
        travel={mockTrip}
        view="grid"
        onClick={() => {}}
        cardLayoutId="card-test-id"
        imageLayoutId="image-test-id"
        isSelected={false}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
