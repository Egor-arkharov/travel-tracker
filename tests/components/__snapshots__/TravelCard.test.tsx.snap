// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TravelCard matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="card grid"
    data-mode="card"
  >
    <div
      class="inner"
    >
      <div
        data-testid="mock-image"
      />
      <div
        data-testid="mock-info"
      />
    </div>
  </article>
</DocumentFragment>
`;
